File: src/plugin/history/History.java
package plugin.history;

import arc.math.geom.Point2;
import arc.struct.LongMap;
import mindustry.net.Administration;
import mindustry.type.UnitType;
import mindustry.world.Block;
import mindustry.world.Tile;

import java.util.Optional;

public class History {
    public static final LongMap<HistoryStack> history = new LongMap<>();

    public static String getMessage(int pos) {
        StringBuilder sb = new StringBuilder();
        // Point2 pos2 = Point2.unpack(pos);
        int x = Point2.x(pos), y = Point2.y(pos); // int int

        sb.append("[" + x + "] " + "[" + y + "]");


        HistoryStack stack = history.get(pos); // long
        //stack.stack.each(s->{sb.append("\n").append(s.getMessage());});
        if (stack != null)
            for (int i = 0; i < stack.size(); i++) {
                sb.append("\n").append(stack.stack.get(i).getMessage());
            }

        return sb.toString();
    }

    public static void write(Tile tile,
                             String playerName,
                             Optional<Integer> playerId,
                             HistoryType type,
                             Block block,
                             UnitType unit) {

        if (tile == null) return;

        HistoryRecord record = new HistoryRecord(playerName, playerId, type, block, unit);

        tile.getLinkedTiles(t -> {

            long pos = t.pos();

            HistoryStack stack = history.get(pos);

            if (stack == null) {
                stack = new HistoryStack();
                history.put(pos, stack);
            }

            if (stack.size() >= 6) {
                stack.removeFirst();
            }

            stack.add(record);

        });
    }

    public static void clear() {
        history.clear();
    }
}

File: src/plugin/history/History.kt
package plugin.history

import arc.math.geom.Point2
import arc.struct.LongMap
import mindustry.game.Team
import mindustry.type.UnitType
import mindustry.world.Block
import mindustry.world.Tile

object History {
    val history: LongMap<HistoryStack> = LongMap()

    private val actors = HashMap<Int, HistoryActor>()

    fun getMessage(pos: Int): String {
        val sb = StringBuilder()

        val x = Point2.x(pos).toInt()
        val y = Point2.y(pos).toInt() // int int

        sb.append("[$x] [$y]")

        synchronized(history) {
            val stack = history.get(pos.toLong()) // long

            if (stack != null) {
                for (i in 0..<stack.size()) {
                    sb.append("\n").append(stack.stack.get(i).getMessage())
                }
            }
        }

        return sb.toString()
    }

    fun copy(): LongMap<HistoryStack> {
        return synchronized(history) {
            val copyMap = LongMap<HistoryStack>(history.size)

            history.forEach { entry ->
                val originalStack = entry.value

                if (originalStack != null) {
                    val stackCopy = HistoryStack()
                    stackCopy.stack.addAll(originalStack.stack)
                    copyMap.put(entry.key, stackCopy)
                }
            }

            copyMap
        }
    }

    fun write(
        tile: Tile?,
        playerName: String?,
        playerId: Int?,
        type: HistoryType,
        block: Block,
        unit: UnitType?,
        team: Team,
        rotation: Int,
        configAfter: Any? = null,
    ) {
        if (tile == null) return

        val time = System.currentTimeMillis()

        val actor = getActor(playerName, playerId)

        val record = HistoryRecord(
            actor,
            type,
            block,
            unit,
            time,
            false,
            team,
            rotation,
            configAfter
        )

        tile.getLinkedTiles { t ->
            val pos = t.pos().toLong()

            if (t.isCenter) {
                addTile(
                    pos,
                    record.copy(center = true)
                )
            } else {
                addTile(pos, record)
            }
        }
    }

    private fun getActor(name: String?, id: Int?): HistoryActor? {
        if (name == null && id == null) return null

        if (id == null) {
            return HistoryActor(name, null)
        }

        return actors.getOrPut(id) {
            HistoryActor(name, id)
        }
    }

    private fun addTile(pos: Long, record: HistoryRecord) {
        synchronized(history) {
            var stack = history.get(pos)

            if (stack == null) {
                stack = HistoryStack()
                history.put(pos, stack)
            }

            if (stack.size() >= 6) {
                stack.removeFirst()
            }

            stack.add(record)
        }
    }

    fun clear() {
        synchronized(history) {
            history.clear()
            actors.clear()
        }
    }
}